### PR TITLE
Implement parallel, batched trial decryption in wallet scanning.

### PR DESCRIPTION
**Pull request by nuttycom**
This ports/reimplements the batched decryption approach used in zcashd to implement light wallet scanning.

### CODE REVIEW

File: zcash_client_backend/src/scan.rs [hunk 1]
```diff
@@ -0,0 +1,221 @@
+use crossbeam_channel as channel;
```
Comment (daira) [inline zcash_client_backend/src/scan.rs:1]: This isn't a standard library import; shouldn't it go in a different section?

File: zcash_client_backend/src/welding_rig.rs [hunk 2]
```diff
@@ -141,16 +144,65 @@ impl ScanningKey for SaplingIvk {
...
+        tree,
+        existing_witnesses,
+        None,
+    )
+}
+
+pub(crate) fn add_block_to_runner<P: consensus::Parameters + Send + 'static>(
+    params: &P,
+    block: CompactBlock,
+    batch_runner: &mut BatchRunner<SaplingDomain<P>, CompactOutputDescription>,
```
Comment (daira) [inline zcash_client_backend/src/welding_rig.rs:169]: How hard will it be to adapt this to cover Orchard batch scanning? I'm concerned that we could end up duplicating quite a lot of code and I want to minimise that as far as possible.

Comment (str4d) [reply zcash_client_backend/src/welding_rig.rs:169]: I've already started the process of adding Orchard to the zcashd version of this in https://github.com/zcash/zcash/pull/6105. Once that is done I'll apply the same changes to this code, assuming it still exists (and hasn't been replaced with async logic).

File: zcash_client_backend/src/scan.rs (hunk 1 above)
Comment (nuttycom) [reply zcash_client_backend/src/scan.rs:1]: We generally group imports from all "external" libraries together.

Comment (str4d) [reply zcash_client_backend/src/scan.rs:1]: I do try and group `std / core / alloc` imports separately from external crate dependencies, but it's easy for an editor doing automatic imports to not pay attention to that sometimes.

File: zcash_client_backend/src/scan.rs [hunk 3]
```diff
@@ -1,200 +1,19 @@
-use core::fmt;
+use crossbeam_channel as channel;
 use std::collections::HashMap;
-use std::io;
+use std::fmt;
 use std::mem;
 
-use crossbeam_channel as channel;
```
Comment (str4d) [inline zcash_client_backend/src/scan.rs]: To @daira's point, this change in the second commit would not be present if the split between `core/std` imports and external crate imports had been preserved.

File: zcash_client_backend/src/scan.rs [hunk 4]
```diff
@@ -1,200 +1,19 @@
...
     /// The incoming viewing key used to decrypt the note.
-    ivk: D::IncomingViewingKey,
+    pub(crate) ivk: D::IncomingViewingKey,
     /// The recipient of the note.
-    recipient: D::Recipient,
+    pub(crate) recipient: D::Recipient,
     /// The note!
-    note: D::Note,
-    /// The memo sent with the note.
-    memo: D::Memo,
```
Comment (str4d) [inline zcash_client_backend/src/scan.rs]: Non-blocking: we should figure out how to generalise this logic over compact and non-compact outputs, so we can potentially share it with `zcashd` (assuming the logic doesn't change significantly when we move to async).

**Review by str4d — APPROVED** [review]
utACK 397c76ca8d3972855392b7a3a4ec975a6958f674

File: zcash_client_backend/src/scan.rs (hunk 4 above)
Comment (nuttycom) [reply zcash_client_backend/src/scan.rs]: Agreed; I made a quick attempt at it before we started pairing on this but it added enough complexity that think that should be a separate refactoring.

**Pull request merged**